### PR TITLE
Refactor StorageResource state in base class.

### PR DESCRIPTION
**Pull request by pshapiro4broad**
Move all state into StorageResource base class; remove setters and make fields final.

I thought I would give this a try and see how it went. I didn't run the tests but the IDE seemed OK with it.

### CODE REVIEW

File: src/main/java/bio/terra/service/dataset/StorageResource.java [hunk 1]
```diff
@@ -2,39 +2,74 @@
...
+import java.util.Objects;
 import java.util.UUID;
 
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.EXISTING_PROPERTY, property = "cloudPlatform")
 @JsonSubTypes({
     @JsonSubTypes.Type(value = AzureStorageResource.class, name = "azure"),
     @JsonSubTypes.Type(value = GoogleStorageResource.class, name = "gcp")
 })
-public interface StorageResource<Region extends CloudRegion, Resource extends CloudResource> {
+public abstract class StorageResource<Region extends CloudRegion, Resource extends CloudResource> {
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/StorageResource.java:18]: I thought you were trying to do away with the generics?

Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/StorageResource.java:18]: I think I'm close to getting a version without the generics to work

Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/dataset/StorageResource.java:18]: Well, I was trying it out and it didn't seem so bad. I like that the subclasses are pretty much marker classes with this change.

Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/StorageResource.java:18]: Welp, nevermind...couldn't get it to work without the generics...No a huge deal, just means that we need a bunch of `<?, ?>`'s in the code when we have `StorageResource` variables

**Pull request closed without merging**